src/ipc/ipc4/helper.c

```diff
@@ -1002,11 +1003,44 @@ int ipc4_find_dma_config(struct ipc_config_dai *dai, uint8_t *data_buffer, uint3
 	if (*dma_config_id != GTW_DMA_CONFIG_ID)
 		return IPC4_INVALID_REQUEST;
 
-	dai->host_dma_config = GET_IPC_DMA_CONFIG(data_buffer, size);
+	dai->host_dma_config[0] = GET_IPC_DMA_CONFIG(data_buffer, size);
 #endif
 	return IPC4_SUCCESS;
 }
 
+int ipc4_find_dma_config_multiple(struct ipc_config_dai *dai, uint8_t *data_buffer,
+				  uint32_t size, uint32_t device_id)
+{
+	uint32_t end_addr = (uint32_t)data_buffer + size;
+	struct ipc_dma_config *dma_cfg;
+	struct sof_tlv *tlvs;
+	int dma_cfg_count = 0;
+
+	for (tlvs = (struct sof_tlv *)data_buffer; (uint32_t)tlvs < end_addr;
+	     tlvs = tlv_next(tlvs)) {
+
+		dma_cfg = tlv_value_ptr_get(tlvs, GTW_DMA_CONFIG_ID);
+		if (!dma_cfg)
+			continue;
+
+		if (dma_cfg->channel_map.device_count == 0) {
+			dai->host_dma_config[dma_cfg_count] = dma_cfg;
+			return IPC4_SUCCESS;
+		}
+
+		for (uint32_t i = 0; i < dma_cfg->channel_map.device_count; i++) {
+			if (dma_cfg->channel_map.map[i].device_address == device_id) {
+				dai->host_dma_config[dma_cfg_count] = dma_cfg;
+				return IPC4_SUCCESS;
+			}
+		}
+
+		dma_cfg_count++;
```

Comment:
@iganakov We should only do `dma_cfg_count++` when we assign `dai->host_dma_config[dma_cfg_count] = dma_cfg;`.
IMHO, dma_cfg_count should be an argument of `ipc4_find_dma_config_multiple` not a local variable. `ipc4_find_dma_config_multiple` will return immediately when it find the dma config with the device_id. We need to specify dai->host_dma_config[] index from the caller.
I do meet an issue if the dma config is not found in the first tlv. In that case,  we don't assign `dai->host_dma_config[dma_cfg_count] = dma_cfg;` in the first `for (uint32_t i = 0; i < dma_cfg->channel_map.device_count; i++)` loop, and `dma_cfg_count++`. As a result, `dai->host_dma_config[0]` = NULL and dma_cfg is assigned to `dai->host_dma_config[1]` that is not what we want.

Reply:
Good catch. I'll change `ipc4_find_dma_config_multiple()` in next revision of this patch. Thanks @bardliao